File: backend/src/main/java/com/tailorstudio/app/web/BusinessController.java
package com.tailorstudio.app.web;

import com.tailorstudio.app.domain.UserRole;
import com.tailorstudio.app.dto.BusinessUpdateRequest;
import com.tailorstudio.app.repo.BusinessRepository;
import com.tailorstudio.app.security.CurrentUserService;
import com.tailorstudio.app.service.AuthService;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

@RestController
@RequestMapping("/api/business")
public class BusinessController {

    private final CurrentUserService currentUserService;
    private final BusinessRepository businessRepository;
    private final AuthService authService;

    public BusinessController(
            CurrentUserService currentUserService,
            BusinessRepository businessRepository,
            AuthService authService) {
        this.currentUserService = currentUserService;
        this.businessRepository = businessRepository;
        this.authService = authService;
    }

    @PreAuthorize("hasRole('OWNER')")
    @PatchMapping
    public Map<String, String> update(@Valid @RequestBody BusinessUpdateRequest req) {
        var u = currentUserService.requireUser();
        var b = businessRepository.findById(u.getBusinessId()).orElseThrow();
        if (req.name() != null && !req.name().isBlank()) {
            b.setName(req.name().trim());
        }
        if (req.tagline() != null) {
            b.setTagline(req.tagline().isBlank() ? null : req.tagline().trim());
        }
        if (req.address() != null) {
            b.setAddress(req.address().isBlank() ? null : req.address().trim());
        }
        if (req.phone() != null) {
            b.setPhone(req.phone().isBlank() ? null : req.phone().trim());
        }
        if (req.secondaryPhone() != null) {
            b.setSecondaryPhone(req.secondaryPhone().isBlank() ? null : req.secondaryPhone().trim());
        }
        businessRepository.save(b);
        return Map.of("ok", "true");
    }

    @PreAuthorize("hasRole('OWNER')")
    @PostMapping("/rotate-join-code")
    public Map<String, String> rotateJoinCode() {
        var u = currentUserService.requireUser();
        if (u.getRole() != UserRole.OWNER) {
            throw new IllegalArgumentException("Only owners can rotate the code");
        }
        String code = authService.rotateJoinCode(u.getBusinessId());
        return Map.of("joinCode", code);
    }
}

File: server/src/controllers/businessController.ts
import type { Response } from 'express'
import type { AuthedRequest } from '../middleware/auth.js'
import { Business } from '../models/Business.js'
import { rotateJoinCode } from '../services/authService.js'
import { parseObjectId } from '../utils/objectId.js'

export async function patchBusiness(req: AuthedRequest, res: Response) {
  const b = await Business.findById(parseObjectId(req.user.businessId))
  if (!b) {
    res.status(404).json({ error: 'Not found' })
    return
  }
  const body = req.body as Record<string, string | undefined>
  if (body.name?.trim()) b.name = body.name.trim()
  if (body.tagline !== undefined) b.tagline = body.tagline?.trim() || null
  if (body.address !== undefined) b.address = body.address?.trim() || null
  if (body.phone !== undefined) b.phone = body.phone?.trim() || null
  if (body.secondaryPhone !== undefined) b.secondaryPhone = body.secondaryPhone?.trim() || null
  await b.save()
  res.json({ ok: 'true' })
}

export async function rotateCode(req: AuthedRequest, res: Response) {
  if (req.user.role !== 'OWNER') {
    res.status(400).json({ error: 'Only owners can rotate the code' })
    return
  }
  const joinCode = await rotateJoinCode(req.user.businessId)
  res.json({ joinCode })
}
